### DISCUSSION TITLE
[Feature Request] Fabric Markup Language

### DISCUSSION THREAD

**Discussion by Ayehavgunne** [Fabric-Development/fabric#45] [Ideas]
When looking at the examples I immediately noticed the unsightly amount of nesting. I then had a thought. It wouldn't be too hard to have a little dsl like JSX does with HTML. A simple parser could take some xml like:

```xml
<Box spacing=24 name="main-window" orientation="v">
    <Box spacing=14 name="header" orientation="h">
        ...
    </Box>
</Box>
```
then transform it to the equivalent python code.

```python
Box(
    spacing=24,
    name="main-window",
    orientation="v",
    children=[
        Box(
            spacing=14,
            name="header",
            orientation="h",
            ...
```

Then this markup could be stored in a separate .fml file and referenced like the css is.

I _might_ even be able to help out code such a feature if needed.

**Comment by its-darsh** [comment]
Fabric is intended to be a Python package that offers a strong DevEX. It utilizes the LSP of your code editor to provide you with a type-hinted session.

I don't believe that Fabric was ever intended to be used in such a way. Moreover, reimplementing this concept in Python might not be the best approach - it could be slow, difficult to implement, and challenging to maintain.

I recommend looking into [Gtk.Builder](https://docs.gtk.org/gtk3/class.Builder) as it provides functionality similar to what you're aiming for.

If you're still into the idea of writing a DSL for Fabric, it's definitely possible to do so and release it as an extension for Fabric. You could also share your code as a snippet on the "snippets" page of the [new wiki](https://github.com/its-darsh/fabric-wiki).
